File: LeetCode/0447. Number of Boomeranges/Solution.kt
/**
 * 31 / 31 test cases passed.
 * Status: Accepted
 * Runtime: 938 ms
 */
class Solution {
    fun numberOfBoomerangs(points: Array<IntArray>): Int {
        var res = 0
        points.forEachIndexed { i, p1 ->
            val store = HashMap<Int, Int>()
                if (i != j) {
                    val dis = dis(p1, p2)
                    store[dis] = (store[dis] ?: 0) + 1
                }
            }
            store.forEach { _, v ->
                res += v * (v - 1)
            }
        }
        return res
    }

    private fun dis(p1: IntArray, p2: IntArray)
            : Int = (p1[0] - p2[0]) * (p1[0] - p2[0]) + (p1[1] - p2[1]) * (p1[1] - p2[1])
}

fun main(args: Array<String>) {
    Solution().numberOfBoomerangs(arrayOf(
            intArrayOf(0, 0),
            intArrayOf(1, 0),
            intArrayOf(2, 0))).let(::println)
}

File: LeetCode/0447. Number of Boomeranges/solution.cpp
/**
 * 32 / 32 test cases passed.
 * Runtime: 476 ms
 * Memory Usage: 121.2 MB 
 */
class Solution {
public:
    int distance(const vector<int>& p1, const vector<int>& p2) {
        return pow(p1[0] - p2[0], 2) + pow(p1[1] - p2[1], 2);
    }
    int numberOfBoomerangs(vector<vector<int>>& points) { 
        int ans = 0;
        for (const vector<int>& p: points) {
            unordered_map<int, int> countit;
            for (const vector<int>& q: points) { 
                int dist = distance(p, q);
                countit[dist]++;
            }
            for (auto& [_, val]: countit) {
                ans += val * (val - 1);
            }
        }
        return ans;
    }
};
